# 📊 Exploratory Data Analysis (EDA)
## Dataset: Student Performance Factors

Este proyecto analiza factores que influyen en el rendimiento estudiantil usando un dataset público de Kaggle.  
El objetivo es aplicar técnicas de **limpieza de datos**, **exploración** y **visualización** para extraer patrones relevantes.

## 🔧 1. Carga de librerías y dataset
En esta sección importamos las librerías necesarias y cargamos el dataset desde la carpeta `../data/`.


In [1]:
# Importar librerías
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
import warnings
warnings.filterwarnings('ignore')

In [23]:
# Cargar el dataset
df = pd.read_csv("../data/student_performance.csv")

## 🔍 2. Exploración inicial
Aquí revisamos la forma del dataset, sus columnas y tipos de datos.


In [24]:
# Exploración inicial
print("Forma del dataset:", df.shape) # Muestra filas x columnas
print("\nColumnas:")
print(df.columns)

df.head()

Forma del dataset: (6607, 20)

Columnas:
Index(['Hours_Studied', 'Attendance', 'Parental_Involvement',
       'Access_to_Resources', 'Extracurricular_Activities', 'Sleep_Hours',
       'Previous_Scores', 'Motivation_Level', 'Internet_Access',
       'Tutoring_Sessions', 'Family_Income', 'Teacher_Quality', 'School_Type',
       'Peer_Influence', 'Physical_Activity', 'Learning_Disabilities',
       'Parental_Education_Level', 'Distance_from_Home', 'Gender',
       'Exam_Score'],
      dtype='object')


Hours_Studied  Attendance Parental_Involvement Access_to_Resources  \
0             23          84                  Low                High   
1             19          64                  Low              Medium   
2             24          98               Medium              Medium   
3             29          89                  Low              Medium   
4             19          92               Medium              Medium   

  Extracurricular_Activities  Sleep_Hours  Previous_Scores Motivation_Level  \
0                         No            7               73              Low   
1                         No            8               59              Low   
2                        Yes            7               91           Medium   
3                        Yes            8               98           Medium   
4                        Yes            6               65           Medium   

  Internet_Access  Tutoring_Sessions Family_Income Teacher_Quality  \
0             Yes                  0           Low          Medium   
1             Yes                  2        Medium          Medium   
2             Yes                  2        Medium          Medium   
3             Yes                  1        Medium          Medium   
4             Yes                  3        Medium            High   

  School_Type Peer_Influence  Physical_Activity Learning_Disabilities  \
0      Public       Positive                  3                    No   
1      Public       Negative                  4                    No   
2      Public        Neutral                  4                    No   
3      Public       Negative                  4                    No   
4      Public        Neutral                  4                    No   

  Parental_Education_Level Distance_from_Home  Gender  Exam_Score  
0              High School               Near    Male          67  
1                  College           Moderate  Female          61  
2             Postgraduate               Near    Male          74  
3              High School           Moderate    Male          71  
4                  College               Near  Female          70

In [25]:
# Información general
df.info()

# Estadísticas descriptivas
df.describe(include='all') # Con include='all' hacemos que lo haga con cada columna


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6607 entries, 0 to 6606
Data columns (total 20 columns):
 #   Column                      Non-Null Count  Dtype 
---  ------                      --------------  ----- 
 0   Hours_Studied               6607 non-null   int64 
 1   Attendance                  6607 non-null   int64 
 2   Parental_Involvement        6607 non-null   object
 3   Access_to_Resources         6607 non-null   object
 4   Extracurricular_Activities  6607 non-null   object
 5   Sleep_Hours                 6607 non-null   int64 
 6   Previous_Scores             6607 non-null   int64 
 7   Motivation_Level            6607 non-null   object
 8   Internet_Access             6607 non-null   object
 9   Tutoring_Sessions           6607 non-null   int64 
 10  Family_Income               6607 non-null   object
 11  Teacher_Quality             6529 non-null   object
 12  School_Type                 6607 non-null   object
 13  Peer_Influence              6607 non-null   obje

Hours_Studied   Attendance Parental_Involvement Access_to_Resources  \
count     6607.000000  6607.000000                 6607                6607   
unique            NaN          NaN                    3                   3   
top               NaN          NaN               Medium              Medium   
freq              NaN          NaN                 3362                3319   
mean        19.975329    79.977448                  NaN                 NaN   
std          5.990594    11.547475                  NaN                 NaN   
min          1.000000    60.000000                  NaN                 NaN   
25%         16.000000    70.000000                  NaN                 NaN   
50%         20.000000    80.000000                  NaN                 NaN   
75%         24.000000    90.000000                  NaN                 NaN   
max         44.000000   100.000000                  NaN                 NaN   

       Extracurricular_Activities  Sleep_Hours  Previous_Scores  \
count                        6607   6607.00000      6607.000000   
unique                          2          NaN              NaN   
top                           Yes          NaN              NaN   
freq                         3938          NaN              NaN   
mean                          NaN      7.02906        75.070531   
std                           NaN      1.46812        14.399784   
min                           NaN      4.00000        50.000000   
25%                           NaN      6.00000        63.000000   
50%                           NaN      7.00000        75.000000   
75%                           NaN      8.00000        88.000000   
max                           NaN     10.00000       100.000000   

       Motivation_Level Internet_Access  Tutoring_Sessions Family_Income  \
count              6607            6607        6607.000000          6607   
unique                3               2                NaN             3   
top              Medium             Yes                NaN           Low   
freq               3351            6108                NaN          2672   
mean                NaN             NaN           1.493719           NaN   
std                 NaN             NaN           1.230570           NaN   
min                 NaN             NaN           0.000000           NaN   
25%                 NaN             NaN           1.000000           NaN   
50%                 NaN             NaN           1.000000           NaN   
75%                 NaN             NaN           2.000000           NaN   
max                 NaN             NaN           8.000000           NaN   

       Teacher_Quality School_Type Peer_Influence  Physical_Activity  \
count             6529        6607           6607        6607.000000   
unique               3           2              3                NaN   
top             Medium      Public       Positive                NaN   
freq              3925        4598           2638                NaN   
mean               NaN         NaN            NaN           2.967610   
std                NaN         NaN            NaN           1.031231   
min                NaN         NaN            NaN           0.000000   
25%                NaN         NaN            NaN           2.000000   
50%                NaN         NaN            NaN           3.000000   
75%                NaN         NaN            NaN           4.000000   
max                NaN         NaN            NaN           6.000000   

       Learning_Disabilities Parental_Education_Level Distance_from_Home  \
count                   6607                     6517               6540   
unique                     2                        3                  3   
top                       No              High School               Near   
freq                    5912                     3223               3884   
mean                     NaN                      NaN                NaN   
std                 

## 📋 Observaciones iniciales

- El dataset contiene **6607 filas** y **20 columnas**.  
- Las columnas incluyen tanto **variables numéricas** (ej. `Exam_Score`, `Hours_Studied`) como **categóricas** (ej. `Gender`, `Parental_Involvement`, `Motivation_Level`).  
- Podemos observar posibles **valores nulos** en algunas columnas, que deberán ser tratados en la fase de limpieza.  
- Hay variables con valores de texto que podrían necesitar **normalización** (ej. categorías con mayúsculas/minúsculas o espacios).  
- Los nombres de las columnas ya estaban en `snake_case`, por lo que solo se convirtieron a **minúsculas**.
- El dataset parece adecuado para un análisis exploratorio, ya que combina factores sociales, académicos y personales relacionados con el rendimiento estudiantil.


## 🧹 3. Limpieza de datos

En esta sección se realiza una limpieza básica del dataset para asegurar su calidad antes del análisis.  
Se revisan duplicados, valores nulos, tipos de datos y posibles normalizaciones.


### Normalización de nombres de columnas

Para facilitar el acceso a las columnas y mantener la coherencia del código,  
se van a convertir todos los nombres de las columnas a **minúsculas**.

In [26]:
# Convertir nombres de columnas a minúsculas
#df.columns = df.columns.str.lower()
df.columns = [i.lower() for i in df.columns]
df

hours_studied  attendance parental_involvement access_to_resources  \
0                23          84                  Low                High   
1                19          64                  Low              Medium   
2                24          98               Medium              Medium   
3                29          89                  Low              Medium   
4                19          92               Medium              Medium   
...             ...         ...                  ...                 ...   
6602             25          69                 High              Medium   
6603             23          76                 High              Medium   
6604             20          90               Medium                 Low   
6605             10          86                 High                High   
6606             15          67               Medium                 Low   

     extracurricular_activities  sleep_hours  previous_scores  \
0                            No            7               73   
1                            No            8               59   
2                           Yes            7               91   
3                           Yes            8               98   
4                           Yes            6               65   
...                         ...          ...              ...   
6602                         No            7               76   
6603                         No            8               81   
6604                        Yes            6               65   
6605                        Yes            6               91   
6606                        Yes            9               94   

     motivation_level internet_access  tutoring_sessions family_income  \
0                 Low             Yes                  0           Low   
1                 Low             Yes                  2        Medium   
2              Medium             Yes                  2        Medium   
3              Medium             Yes                  1        Medium   
4              Medium             Yes                  3        Medium   
...               ...             ...                ...           ...   
6602           Medium             Yes                  1          High   
6603           Medium             Yes                  3           Low   
6604              Low             Yes                  3           Low   
6605             High             Yes                  2           Low   
6606           Medium             Yes                  0        Medium   

     teacher_quality school_type peer_influence  physical_activity  \
0             Medium      Public       Positive                  3   
1             Medium      Public       Negative                  4   
2             Medium      Public        Neutral                  4   
3             Medium      Public       Negative                  4   
4               High      Public        Neutral                  4   
...              ...         ...            ...                ...   
6602          Medium      Public       Positive                  2   
6603            High      Public       Positive                  2   
6604          Medium      Public       Negative                  2   
6605          Medium     Private       Positive                  3   
6606          Medium      Public       Positive                  4   

     learning_disabilities parental_education_level distance_from_home  \
0                       No              High School               Near   
1                       No                  College           Moderate   
2                       No             Postgraduate               Near   
3                       No              High School           Moderate   
4                       No                  College               Near   
...                    ...                      ...                ...   
6602                    No              High School             

### Duplicados

In [27]:
# Comprobación de duplicados
duplicados = df.duplicated().sum()
print(f"Número de filas duplicadas: {duplicados}")


Número de filas duplicadas: 0


Se realizó una comprobación de duplicados en el dataset.  
El resultado fue **0 filas duplicadas**, por lo tanto **no fue necesario eliminar registros repetidos**.  


### Valores Nulos

In [28]:
# Comprobación de valores nulos
df.isna().sum()

hours_studied                  0
attendance                     0
parental_involvement           0
access_to_resources            0
extracurricular_activities     0
sleep_hours                    0
previous_scores                0
motivation_level               0
internet_access                0
tutoring_sessions              0
family_income                  0
teacher_quality               78
school_type                    0
peer_influence                 0
physical_activity              0
learning_disabilities          0
parental_education_level      90
distance_from_home            67
gender                         0
exam_score                     0
dtype: int64

Se detectaron valores *nulos* en **tres columnas** del dataset:

- `Teacher_Quality` (78 nulos): **categórica**
- `Parental_Education_Level` (90 nulos): **categórica**
- `Distance_from_Home` (67 nulos): **categórica**

Todas las columnas serán tratadas con **sustitución** por **modo**, ya que son de tipo categórico.  
Esta estrategia preserva la coherencia del análisis y evita introducir sesgos numéricos.

In [ ]:
# Sustitución mode()[0] máxima frecuencia
df["teacher_quality"].fillna(df["teacher_quality"].mode()[0], inplace=True)
df["distance_from_home"].fillna(df["distance_from_home"].mode()[0], inplace=True)
df["parental_education_level"].fillna(df["parental_education_level"].mode()[0], inplace=True)

### Tipos de datos

En el dataset, se identificaron 13 columnas categóricas con valores como `"Yes/No"`, `"Low/Medium/High"` o `"Public/Private"`.  
Estas columnas se convierten al tipo `category` para:

- Optimizar el uso de memoria
- Facilitar visualizaciones y agrupaciones

Columnas convertidas:

- `parental_involvement`
- `access_to_resources`
- `extracurricular_activities`
- `motivation_level`
- `internet_access`
- `family_income`
- `teacher_quality`
- `school_type`
- `peer_influence`
- `learning_disabilities`
- `parental_education_level`
- `distance_from_home`
- `gender`

In [31]:
# Conversión de columnas categóricas a tipo 'category'
categoricas = [
    "parental_involvement",
    "access_to_resources",
    "extracurricular_activities",
    "motivation_level",
    "internet_access",
    "family_income",
    "teacher_quality",
    "school_type",
    "peer_influence",
    "learning_disabilities",
    "parental_education_level",
    "distance_from_home",
    "gender"
]
df[categoricas] = df[categoricas].astype("category")

In [ ]:
# Verificamos con info() que se realizó el cambio
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6607 entries, 0 to 6606
Data columns (total 20 columns):
 #   Column                      Non-Null Count  Dtype   
---  ------                      --------------  -----   
 0   hours_studied               6607 non-null   int64   
 1   attendance                  6607 non-null   int64   
 2   parental_involvement        6607 non-null   category
 3   access_to_resources         6607 non-null   category
 4   extracurricular_activities  6607 non-null   category
 5   sleep_hours                 6607 non-null   int64   
 6   previous_scores             6607 non-null   int64   
 7   motivation_level            6607 non-null   category
 8   internet_access             6607 non-null   category
 9   tutoring_sessions           6607 non-null   int64   
 10  family_income               6607 non-null   category
 11  teacher_quality             6607 non-null   category
 12  school_type                 6607 non-null   category
 13  peer_influence    

### Revisión de rangos y coherencia

Además de la limpieza básica (duplicados, nulos, nombres de columnas y tipos de datos),  
es importante verificar que los valores del dataset sean **coherentes y estén dentro de rangos lógicos**.  

En esta sección se revisará:

- Que las variables numéricas (`attendance`, `sleep_hours`, `hours_studied`, `tutoring_sessions`, `physical_activity`, `exam_score`, `previous_scores`) no tengan valores fuera de rango o negativos.
- Que las variables categóricas (`gender`, `school_type`, `teacher_quality`, etc.) contengan únicamente las categorías definidas en la documentación oficial.
- Que la variable objetivo (`exam_score`) no presente valores nulos y tenga una distribución razonable.




In [34]:
# Revisar rangos en columnas numéricas
print("% Asistencia fuera de rango:", df[~df["attendance"].between(0, 100)].shape[0])
print("Horas de sueño fuera de rango:", df[~df["sleep_hours"].between(0, 24)].shape[0])
print("Horas de estudio negativas:", df[df["hours_studied"] < 0].shape[0])
print("Sesiones de tutoría negativas:", df[df["tutoring_sessions"] < 0].shape[0])
print("Actividad física semanal negativas:", df[df["physical_activity"] < 0].shape[0])

# Revisar categorías únicas
for col in df.select_dtypes(include="category").columns:
    print(f"{col}: {df[col].unique()}")

% Asistencia fuera de rango: 0
Horas de sueño fuera de rango: 0
Horas de estudio negativas: 0
Sesiones de tutoría negativas: 0
Actividad física semanal negativas: 0
parental_involvement: ['Low', 'Medium', 'High']
Categories (3, object): ['High', 'Low', 'Medium']
access_to_resources: ['High', 'Medium', 'Low']
Categories (3, object): ['High', 'Low', 'Medium']
extracurricular_activities: ['No', 'Yes']
Categories (2, object): ['No', 'Yes']
motivation_level: ['Low', 'Medium', 'High']
Categories (3, object): ['High', 'Low', 'Medium']
internet_access: ['Yes', 'No']
Categories (2, object): ['No', 'Yes']
family_income: ['Low', 'Medium', 'High']
Categories (3, object): ['High', 'Low', 'Medium']
teacher_quality: ['Medium', 'High', 'Low']
Categories (3, object): ['High', 'Low', 'Medium']
school_type: ['Public', 'Private']
Categories (2, object): ['Private', 'Public']
peer_influence: ['Positive', 'Negative', 'Neutral']
Categories (3, object): ['Negative', 'Neutral', 'Positive']
learning_disabilitie

Se verificaron los valores de las columnas numéricas y categóricas:

- `attendance` está dentro del rango 0–100.  
- `sleep_hours` no presenta valores fuera del rango 0–24.  
- `hours_studied`, `tutoring_sessions` y `physical_activity` no tienen valores negativos.  
- Las variables categóricas (`gender`, `school_type`, `teacher_quality`, etc.) presentan únicamente las categorías definidas en la documentación oficial.

### ✅ Limpieza completada

Se ha realizado una limpieza completa del dataset, incluyendo:

- Eliminación de duplicados (no se encontraron registros repetidos)
- Sustitución de valores nulos en columnas categóricas usando el modo
- Normalización de nombres de columnas a minúsculas
- Conversión de columnas categóricas al tipo `category` según la documentación oficial
- Verificación de rangos en variables numéricas y coherencia en categorías

Con estos pasos, el dataset queda listo para la fase de visualización y análisis exploratorio.
